android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
 public class TiDatabaseProxy extends KrollProxy
 {
 	private static final String TAG = "TiDB";
 
+	private static Thread thread;
+	private static BlockingQueue<Runnable> queue = new LinkedBlockingQueue<>();
+	private static AtomicBoolean executingQueue = new AtomicBoolean(false);
+
 	protected SQLiteDatabase db;
 	protected String name;
-	boolean statementLogging, readOnly;
 
+	/**
+	 * Create a TiDatabaseProxy for SQLiteDatabase instance.
+	 * @param name Database name.
+	 * @param db SQLiteDatabase instance.
+	 */
 	public TiDatabaseProxy(String name, SQLiteDatabase db)
 	{
 		super();
+
+		if (this.thread == null) {
+
+			// Query execution thread.
+			this.thread = new Thread(new Runnable() {
+				@Override
+				public void run()
+				{
+					try {
+
+						// Query execution loop.
+						while (true) {
+
+							// Obtain query and execute.
+							// NOTE: This will block until a query is available.
+							queue.take().run();
+
+							// Queue empty? Send notify event.
+							if (queue.isEmpty()) {
+								synchronized (executingQueue)
+								{
+									executingQueue.set(false);
+									executingQueue.notify();
+								}
+							}
+						}
+					} catch (InterruptedException e) {
+					}
+				}
+			});
+			this.thread.start();
+		}
+
 		this.name = name;
 		this.db = db;
-		statementLogging = false;
-		readOnly = false;
 	}
 
-	// readonly database
+	/**
+	 * Create a TiDatabaseProxy for SQLiteDatabase instance.
+	 * @param db SQLiteDatabase instance.
+	 */
 	public TiDatabaseProxy(SQLiteDatabase db)
 	{
-		super();
-		this.name = db.getPath();
-		this.db = db;
-		statementLogging = false;
-		readOnly = true;
+		this(db.getPath(), db);
 	}
 
+	/**
+	 * Close database.
+	 */
 	@Kroll.method
 	public void close()
 	{
 		if (db.isOpen()) {
-			Log.d(TAG, "Closing database: " + name, Log.DEBUG_MODE);
 			db.close();
-		} else {
-			Log.d(TAG, "Database is not open, ignoring close for " + name, Log.DEBUG_MODE);
 		}
 	}
 
+	/**
+	 * Synchronously execute a single SQL query.
+	 * @param query SQL query to execute on database.
+	 * @param parameterObjects Parameters for `query`
+	 */
 	@Kroll.method
-	public TiResultSetProxy execute(String sql, Object... args)
+	private TiResultSetProxy execute(String query, Object... parameterObjects)
 	{
-		// Handle the cases where an array is passed containing the SQL query arguments.
-		// Otherwise use the variable argument list for the SQL query.
-		Object[] sqlArgs;
-		if (args != null && args.length == 1 && args[0] instanceof Object[]) {
-			sqlArgs = (Object[]) args[0];
-		} else {
-			sqlArgs = args;
+		// Validate and parse parameter objects.
+		if (parameterObjects != null && parameterObjects.length == 1 && parameterObjects[0] instanceof Object[]) {
+			parameterObjects = (Object[]) parameterObjects[0];
 		}
+		for (int i = 0; i < parameterObjects.length; i++) {
+			if (parameterObjects[i] instanceof TiBlob) {
+				parameterObjects[i] = ((TiBlob) parameterObjects[i]).getBytes();
```

Comment:
This is a behavioral change. We should add some unit tests around this. This is probably the right thing to do for "binary" blobs, but may be wrong and breaking behavior for blobs holding string contents.

Compounding this is the ambiguous case of "file" blobs whose contents may be binary or not.

Reply:
This is actually the same code as what's [here](https://github.com/appcelerator/titanium_mobile/blob/master/android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java#L136). I've just moved it.